File: js/src/services/analysisService.js
/**
 * Analysis service for VNG data - JavaScript version
 */

const { ParsedFile, AnalysisResult, AnalysisResults, MetricData } = typeof module !== 'undefined' && module.exports
    ? require('../domain/models')
    : window.VNGDomain;

const { AnalysisError } = typeof module !== 'undefined' && module.exports
    ? require('../domain/exceptions')
    : window.VNGExceptions;

const { runAnalysis } = typeof module !== 'undefined' && module.exports
    ? require('../modules/analyzer')
    : window.VNGAnalyzer;

/**
 * Service for analyzing VNG data
 */
class AnalysisService {
    /**
     * Analyze multiple parsed files and return AnalysisResults
     * @param {ParsedFile[]} parsedFiles - List of ParsedFile domain models
     * @returns {AnalysisResults} AnalysisResults domain model
     */
    static analyzeFiles(parsedFiles) {
        try {
            if (!Array.isArray(parsedFiles) || parsedFiles.length === 0) {
                return new AnalysisResults({}, 0, 0);
            }

            // Convert to legacy format for existing analyzer
            const fileDataList = parsedFiles.map(pf => ({
                'name': pf.name,
                'data': this._convertParsedFileToDict(pf)
            }));

            // Run analysis
            const rawResults = runAnalysis(fileDataList);

            // Convert to domain models
            const analysisResults = {};
            let totalMetrics = 0;

            for (const [category, metricsMap] of Object.entries(rawResults)) {
                const metricDataDict = {};

                for (const [metric, data] of Object.entries(metricsMap)) {
                    metricDataDict[metric] = new MetricData(
                        data.values,
                        data.flags,
                        data.delta,
                        data.percentChange,
                        data.stdDev
                    );
                    totalMetrics++;
                }

                analysisResults[category] = new AnalysisResult(category, metricDataDict);
            }

            return new AnalysisResults(analysisResults, parsedFiles.length, totalMetrics);
        } catch (error) {
            throw new AnalysisError(`Analysis failed: ${error.message}`);
        }
    }

    /**
     * Convert ParsedFile to dictionary format for legacy analyzer
     * @private
     * @param {ParsedFile} parsedFile - ParsedFile domain model
     * @returns {Object} Dictionary format
     */
    static _convertParsedFileToDict(parsedFile) {
        const result = {};

        for (const [category, metrics] of Object.entries(parsedFile.data)) {
            result[category] = {};

            for (const [metric, metricValue] of Object.entries(metrics)) {
                result[category][metric] = {
                    'value': metricValue.value,
                    'isFlagged': metricValue.isFlagged
                };
            }
        }

        return result;
    }

    /**
     * Analyze files using dictionary format (legacy compatibility)
     * @param {Array} fileDataList - List of file data dictionaries
     * @returns {Object} Dictionary with analysis results
     */
    static analyzeFilesDict(fileDataList) {
        return runAnalysis(fileDataList);
    }
}

// Export service
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { AnalysisService };
} else if (typeof window !== 'undefined') {
    window.VNGServices = window.VNGServices || {};
    window.VNGServices.AnalysisService = AnalysisService;
}

File: _python_backup/services/analysis_service.py
"""
Analysis service for VNG data
"""

from typing import Dict, List, Any
from domain.models import ParsedFile, AnalysisResult, AnalysisResults, MetricData
from domain.exceptions import AnalysisError
from modules.analyzer import run_analysis as _run_analysis
from utils.statistics import calculate_std_dev, calculate_percent_change


class AnalysisService:
    """Service for analyzing VNG data"""
    
    @staticmethod
    def analyze_files(parsed_files: List[ParsedFile]) -> AnalysisResults:
        """
        Analyze multiple parsed files and return AnalysisResults
        
        Args:
            parsed_files: List of ParsedFile domain models
            
        Returns:
            AnalysisResults domain model
            
        Raises:
            AnalysisError: If analysis fails
        """
        try:
            if not parsed_files:
                return AnalysisResults(
                    results={},
                    file_count=0,
                    total_metrics=0
                )
            
            # Convert to legacy format for existing analyzer
            file_data_list = [
                {
                    'name': pf.name,
                    'data': {
                        category: {
                            metric: {
                                'value': mv.value,
                                'is_flagged': mv.is_flagged
                            }
                            for metric, mv in metrics.items()
                        }
                        for category, metrics in pf.data.items()
                    }
                }
                for pf in parsed_files
            ]
            
            # Run analysis
            raw_results = _run_analysis(file_data_list)
            
            # Convert to domain models
            analysis_results: Dict[str, AnalysisResult] = {}
            total_metrics = 0
            
            for category, metrics_map in raw_results.items():
                metric_data_dict: Dict[str, MetricData] = {}
                for metric, data in metrics_map.items():
                    metric_data_dict[metric] = MetricData(
                        values=data['values'],
                        flags=data['flags'],
                        delta=data['delta'],
                        percent_change=data['percent_change'],
                        std_dev=data['std_dev']
                    )
                    total_metrics += 1
                
                analysis_results[category] = AnalysisResult(
                    category=category,
                    metrics=metric_data_dict
                )
            
            return AnalysisResults(
                results=analysis_results,
                file_count=len(parsed_files),
                total_metrics=total_metrics
            )
        except Exception as e:
            raise AnalysisError(f"Analysis failed: {str(e)}") from e
    
    @staticmethod
    def analyze_files_dict(file_data_list: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Analyze files using dictionary format (legacy compatibility)
        
        Args:
            file_data_list: List of file data dictionaries
            
        Returns:
            Dictionary with analysis results
        """
        return _run_analysis(file_data_list)

